File: 56-merge-intervals/merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda x: (x[0], x[1]))
        stack = [intervals[0]]
        for i in range(1, len(intervals)):
            i_start, i_end = intervals[i]
            s_start, s_end = stack[-1]
            if s_start <= i_start <= s_end or i_start <= s_start <= i_end:
                stack.pop()
                stack.append([min(i_start, s_start), max(i_end, s_end)])
            else:
                stack.append(intervals[i])
        return stack

File: 56-merge-intervals/merge-intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
    if (!intervals.length) return []
    intervals.sort((a, b) => a[0] - b[0])
    const stack = []
    for (let i = 0; i < intervals.length; i++) {
        if (i === 0) {
            stack.push(intervals[i])
            continue
        }
        const [sBegin, sEnd] = stack[stack.length - 1]
        const [iBegin, iEnd] = intervals[i]
        const isOverlap = (sBegin <= iBegin && iBegin <= sEnd) || (iBegin <= sBegin && sBegin <= iEnd)
        if (isOverlap) {
            stack.pop()
            stack.push([Math.min(sBegin, iBegin), Math.max(sEnd, iEnd)])
        } else {
            stack.push(intervals[i])
        }
    }
    return stack
};

`
input = [[1,3],[2,6],[8,10],[15,18]]
output =[[1,6],[8,10],[15,18]]
[1,      6   
            8    10
                    15   18
stack = [[1, 6], [8, 10], ]
`